comps/agent/langchain/src/strategy/workflowexec/planner.py

```diff
@@ -0,0 +1,317 @@
+        query = messages[0].content
+        tool_output = messages[-1].content
+
+        response = self.chain.invoke({"input": query, "tool_output": tool_output})
+        return {"messages": [AIMessage(content=response.content)]}
+
+
+class ToolChainNode:
+    """Executes tool calls based on the current state. Returns the tool result in the form of AIMessage.
+
+    Args:
+        state (messages): The current state
+
+    Returns:
+        dict: The updated state with the response appended to messages
+    """
+
+    def __init__(self, tools):
+        self.chain = ToolNode(tools)
+
+    def __call__(self, state):
+        messages = state["messages"]
+        response = self.chain.invoke({"messages": messages})
+
+        sender = state["sender"]
+
+        if sender == "workflow_status_checker":
+            response = json.loads(response["messages"][0].content)
+            return {"messages": [AIMessage(content=response["message"])], "workflow_status": response["status"]}
+
+        else:
+            return {"messages": [AIMessage(content=response["messages"][0].content)]}
+
+
+def should_continue(state: AgentState):
+    messages = state["messages"]
+    last_message = messages[-1]
+    if last_message.tool_calls:
+        return "call_tool"
+    else:
+        return "end"
+
+
+def should_retry(state):
+    num_retry = 0
+
+    for m in state["messages"]:
+        if instruction in m.content:
+            num_retry += 1
+
+    print("**********Num retry: ", num_retry)
+
+    if state["workflow_status"] == "failed":
+        print("Workflow execution failed. Exiting graph")
+        return "end"
+    elif (num_retry < MAX_RETRY) and (state["workflow_status"] == "finished"):
+        return True
+    elif (num_retry < MAX_RETRY) and (not state["workflow_status"] == "finished"):
+        time.sleep(100)  # interval between each status checking retry
+        return False
+    else:
+        print(exceed_retries_message)
+        return "end"
+
+
+class WorkflowExecutorAgentWithLangGraph(BaseAgent):
+    def __init__(self, args, with_memory=False):
+        super().__init__(args)
+
+        workflow_scheduler_node = WorkflowScheduler(self.llm_endpoint, "workflow_scheduler")
+        status_checker_node = WorkflowStatusChecker(self.llm_endpoint, "workflow_status_checker")
+        workflow_retriever_node = WorkflowDataRetriever(self.llm_endpoint, "workflow_data_retriever")
+        reasoning_node = ReasoningNode(self.llm_endpoint)
+        tool_node = ToolChainNode(self.tools_descriptions)
+
+        workflow = StateGraph(AgentState)
+        workflow.add_node("workflow_scheduler", workflow_scheduler_node)
+        workflow.add_node("workflow_status_checker", status_checker_node)
+        workflow.add_node("workflow_data_retriever", workflow_retriever_node)
+        workflow.add_node("reasoning_agent", reasoning_node)
+        workflow.add_node("tools", tool_node)
+        workflow.add_node("status_checker_tool", tool_node)
```

Comment:
tools and status_checker_tool are using same tools, I suggest you either unified all 3 tool_node as 1 or using 3 different name for better workflow control.